controllers/topology/desired_state.go

```diff
@@ -136,15 +137,81 @@ func computeControlPlane(class *clusterTopologyClass, current *clusterTopologySt
 		}
 	}
 
+	// Computes the desired Kubernetes version for the control plane.
+	controlPlaneVersion, err := computeControlPlaneVersion(current)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compute spec.version for the ControlPlane object")
+	}
+
 	// Sets the desired Kubernetes version for the control plane.
-	// TODO: improve this logic by adding support for version upgrade component by component
-	if err := unstructured.SetNestedField(controlPlane.UnstructuredContent(), current.cluster.Spec.Topology.Version, "spec", "version"); err != nil {
+	if err := unstructured.SetNestedField(controlPlane.UnstructuredContent(), controlPlaneVersion, "spec", "version"); err != nil {
 		return nil, errors.Wrap(err, "failed to set spec.version in the ControlPlane object")
 	}
 
 	return controlPlane, nil
 }
 
+// computeControlPlaneVersion computes the ControlPlane version based on the current clusterTopologyState.
+// TODO: we also have to handle the following cases:
+// * ControlPlane.spec.version != ControlPlane.status.version, i.e. ControlPLane rollout is already in progress.
+// * ControlPlane.spec.version != MachineDeployment[].spec.template.spec.version, i.e. one of the MachineDeployments has
+//   a different version then the ControlPlane.
+func computeControlPlaneVersion(current *clusterTopologyState) (string, error) {
+	currentTopologyVersion := current.cluster.Spec.Topology.Version
+
+	// ControlPlane does not exist yet, use currentTopologyVersion for new ControlPlane.
+	if current.controlPlane.object == nil {
+		return currentTopologyVersion, nil
+	}
+
+	currentControlPlaneVersion, err := getControlPlaneVersion(current.controlPlane.object)
+	if err != nil {
+		return "", err
+	}
+
+	// ControlPlane already has the currentTopologyVersion.
+	if currentControlPlaneVersion == currentTopologyVersion {
+		return currentTopologyVersion, nil
+	}
```

Comment:
This return is the exact same as the last one, can we simplify and merge them?

Reply:
I can do that. I thought it might be easier to read and understand if I spell out the different cases explicitly (especially as there will be a few more).